external-crates/move/crates/move-compiler/src/cfgir/optimize/coalesce_locals.rs

```diff
@@ -0,0 +1,576 @@
+use move_symbol_pool::Symbol;
+
+// DEBUG flag for general printing. Defined as a const so the code will be eliminated by the rustc
+// optimizer when debugging is not required.
+const DEBUG_COALESCE: bool = false;
+
+//**************************************************************************************************
+// Entry
+//**************************************************************************************************
+
+/// returns true if anything changed
+pub fn optimize(
+    signature: &FunctionSignature,
+    infinite_loop_starts: &BTreeSet<Label>,
+    locals: UniqueMap<Var, (Mutability, SingleType)>,
+    cfg: &mut MutForwardCFG,
+) -> UniqueMap<Var, (Mutability, SingleType)> {
+    macro_rules! unique_add_or_error {
+        ($step:expr, $map:expr, $var:expr, $mut:expr, $ty:expr) => {{
+            if let Err(_) = $map.add($var, ($mut, $ty)) {
+                eprintln!(
+                    "Error in coalescing locals for local {} at {}",
+                    $var.value(),
+                    $step
+                );
+                return locals;
+            }
+        }};
+    }
+
+    // Get the conlfict graph
+    cfg.recompute(); // Recompute the CFG first because optimizatins may have changed things.
+    let (_final_invariants, per_command_states) = liveness::analyze(cfg, infinite_loop_starts);
+
+    let conflict_graph = conflicts::Graph::new(per_command_states, cfg);
+
+    if DEBUG_COALESCE {
+        println!("\n\ngraph: ");
+        for (entry, conflicts) in conflict_graph.iter() {
+            print!("{} -> ", entry.value());
+            match conflicts {
+                conflicts::Conflicts::Set(vars) => {
+                    let var_string = vars
+                        .iter()
+                        .map(|x| format!("{}", x.value()))
+                        .collect::<Vec<_>>()
+                        .join(", ");
+                    println!("{{{}}}", var_string);
+                }
+                conflicts::Conflicts::All => println!("FULL"),
+            }
+        }
+    }
+
+    let mut new_locals = UniqueMap::new();
+
+    // Collect everything that is going to keep its name. This currently includes:
+    // - All parameters.
+    // - Anything that is borrowed.
+    // - Anything that is not used, but whose type does not have `drop`.
+    let param_set = signature
+        .parameters
+        .iter()
+        .map(|(_, var, _)| var)
+        .collect::<BTreeSet<_>>();
+    let all_set = locals
+        .key_cloned_iter()
+        .filter_map(|(var, (_mut, ty))| {
+            if param_set.contains(&var)
+                || (!(conflict_graph.used(&var)
+                    || ty.value.abilities(ty.loc).has_ability_(Ability_::Drop)))
+                || matches!(
+                    conflict_graph.get_conflicts(&var),
+                    Some(conflicts::Conflicts::All)
+                )
+            {
+                Some(var)
+            } else {
+                None
+            }
```

Comment:
minor nit. I feel like just doing a `filter(...).map` might be more readable here, since the filter function does not naturally create an option